scipy/special/xsf/zeta.h

```diff
@@ -1,10 +1,371 @@
+	    std::complex<double> den = std::pow(k + 1, z);
+	    std::complex<double> term = sign * (zeta_borwein_coeff[k] - 1.0) / den;
+	    result += term;
+	}
+	return result * -1.0/(1.0 - std::pow(2.0, 1.0 - z));
+    }
+
+    /* Compute riemann zeta for complex z and real part >= 0 */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_right_halfplane(std::complex<double> z) {
+	if (z == 1.0) {
+	    return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	}
+	/* Cutoff for using Euler-MacLaurin chosen based on cursory empirical search.
+	 * TODO: Choose cutoffs in a more principled way. */
+	if (z.real() < 50.0 && std::abs(z.imag()) > 50.0) {
+	    if (z.real() >= 0.0 && z.real() < 2.5 && std::abs(z.imag()) > 1e9) {
+		/* Euler-MacLaurin summation starts to take an unreasonable amount of time in this
+		 * region, so just give up and return NaN instead. */
+		set_error("zeta", SF_ERROR_NO_RESULT, NULL);
+		return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	    }
+	    return zeta_euler_maclaurin(z);
+	}
+	return zeta_borwein(z);
+    }
+
+    XSF_HOST_DEVICE inline std::complex<double> exppi(std::complex<double> z) {
+	    // exp(pi*z) for complex z.
+	    double x = z.real();
+	    double y = z.imag();
+	    std::complex<double> factor1(xsf::cospi(y), xsf::sinpi(y));
+	    double factor2 = std::exp(M_PI*x);
+	    return factor1 * factor2;
+	}
+
+    XSF_HOST_DEVICE inline std::complex<double> logsinpi(std::complex<double> z) {
+	/* log(sinpi(z)) using sin(z) = (exp(i*pi*z) - exp(-i*pi*z)) / 2i
+	 *
+	 * No attempt is made to choose any particular branch of the logarithm.
+	 * This is an internal function and the intent is that this that the
+	 * result of log(sinpi(z)) will be added to other terms, and the sum
+	 * will then be exponentiated, making the choice of a specific branch
+	 * unnecessary.
+	 */
+	std::complex<double> result = std::log(xsf::sinpi(z));
+	// If it doesn't overflow, just do the regular calculation.
+	if (std::isfinite(result.real()) && !std::isfinite(result.imag())) {
+	    return result;
+	}
+	/* Otherwise factor before taking log. This is where we may end up
+	 * taking a branch other than the principal branch. */
+	std::complex<double> J(0.0, 1.0);
+	/* Calculating log((exp(i*pi*z) - exp(-i*pi*z)) / 2i). Factor out term
+	 * with larger magnitude before taking log. */
+	if (z.imag() > 0 ) {
+	    /* if z.imag() > 0 then, exp(-i*pi*z) has greatest magnitude. Factor it
+	     * out to get:
+	     * log(exp(-i*pi*z)*((exp(2*i*pi*z) - 1.0)/(2i)) =
+	     * log(exp(-i*pi*z)) + log((exp(2*i*pi*z) - 1.0)/(2i)) =
+	     * -i*pi*z + log((exp(2*i*pi*z) - 1.0)/(2i)) */
+	    return -J * M_PI * z + std::log((exppi(2.0 * z * J) - 1.0) / (2.0*J));
+	}
+	/* if z.imag() < 0 then, exp(i*pi*z) has greatest magnitude. Factor similarly
+	 * to above */
+	return J * M_PI * z + std::log((1.0 - exppi(-2.0 * z * J)) / (2.0*J));
```

Comment:
Is there a `log1p` function we could use? Looks like this could be a source of precision loss near trivial zeros. You fixed it otherwise, but it couldn't hurt if it's available.

Reply:
Yes, `log1p` is in the C++ standard library. I think I'd planned to try that at some point but forgot to.

Reply:
That's fine.